### ISSUE TITLE
New deprecations in Sass 1.79

### ISSUE THREAD

**Issue by mbabker** [twbs/bootstrap#40849]
### Prerequisites

- [X] I have [searched](https://github.com/twbs/bootstrap/issues?utf8=%E2%9C%93&q=is%3Aissue) for duplicate or closed issues
- [X] I have [validated](https://html5.validator.nu/) any HTML to avoid common problems
- [X] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)

### Describe the issue

In Sass 1.79, a number of color-related functions are deprecated, to include `color.red()`, `color.green()`, and `color.blue()`, with the suggested replacement being the new `color.channel()` function.

### Reduced test cases

Compile Bootstrap's SCSS with Sass 1.79

### What operating system(s) are you seeing the problem on?

macOS

### What browser(s) are you seeing the problem on?

Chrome

### What version of Bootstrap are you using?

v5.3.3

**Comment by julien-deramond** [comment]
Hey @mbabker 
Yeah, saw that the other day. Let's keep this issue open for when we'll bump the `sass` dependency in our repo. In the meantime, I'll try to provide a PR to fix the issue.

**Comment by DrowningElysium** [comment]
> Hey @mbabker Yeah, saw that today in our Bootstrap fork at work. Let's keep this issue open for when we'll bump the `sass` dependency in our repo. In the meantime, I'll try to analyze how to fix the issue. Hopefully, it will be possible. Don't hesitate if there's still no PR yet and you have ideas.
> 
> In the meantime, folks having these warnings can pin their `sass` version waiting for a fix on our side (e.g. `npm i sass@1.77.6 --save-exact`).

Someone in [this topic](https://github.com/twbs/bootstrap/issues/40621#issuecomment-2308781735) might have found the solution. It has to do with the order that mixins are used. I myself am getting quite some errors from _type.scss.

**Comment by klunker** [comment]
Deprecation Warning: The legacy JS API is deprecated and will be removed in Dart Sass 2.0.0.
More info: https://sass-lang.com/d/legacy-js-api

**Comment by thet** [comment]
I've stumbled over this issue too and started a PR - however the tests do not yet pass and any help is wanted.

Meanwhile, I will just silence the deprecation warnings as described here:
https://sass-lang.com/documentation/breaking-changes/legacy-js-api/#silencing-warnings

**Comment by ghiscoding** [comment]
> **Edit with really quick analysis**: Using `sass:color` won't probably be possible as Bootstrap 5 still needs to support `node-sass`.

I was quite surprised to see that Bootstrap still support node-sass when it's been tagged as EOL and the repo is archived, though the EOL happened this year, it was however announced a long time ago. Will that be removed only in Bootstrap 6? I don't see any mention of that in the [BS6 project](https://github.com/orgs/twbs/projects/38). I'm also assuming BS6 probably won't come for another couple years, so will node-sass really be supported that long even though it's EOL?

**Comment by julien-deramond** [comment]
> Will that be removed only in Bootstrap 6? I don't see any mention of that in the [BS6 project](https://github.com/orgs/twbs/projects/38). I'm also assuming BS6 probably won't come for another couple years, so will node-sass really be supported that long even though it's EOL?

That's the question indeed. The original plan was to drop `node-sass` support in v6 because it's a breaking change for some users. And yes, v6 is clearly not coming anytime soon, as development hasn't really started yet.

I haven't fully thought through all the options, but here are the ones I'm considering:
1. **Fix the issue with a workaround**, since we can't currently use `sass:color`. This could involve re-implementing an equivalent for the deprecated functions. Of course, there's always a risk that future changes in Sass will render `node-sass` unusable on our side. And clearly, they have perfectly the right to do so, as `node-sass` is EOL, and announced a long time ago.
2. **Delay the fix until the problem is resolved in Bootstrap** (whatever the version). Meanwhile, users would need to either pin their Sass dependency or ignore warning messages. This approach also gives us several options for when to drop `node-sass`:
  2.1. **Drop `node-sass` support in a v5.3.x release.** This would be unfair to some users still relying on it, and frankly, I’m not a big fan of this option.
  2.2. **Drop `node-sass` support in v5.4.** While it would be a breaking change, clear communication could help. However, this also means that any fixes for the v5.3.x series would need to be addressed before removing `node-sass`.
  2.3. **Stick with the original plan to drop `node-sass` in v6.** This has always been the long-term goal, but with v6 far off, it's worth considering if we need a more immediate solution.

For now, I’m inclined to try option 1. If that doesn’t work, we'll need to weigh our alternatives carefully.

**Comment by louismaximepiton** [comment]
Hey there,

I tried my best finding a workaround with Sass but the only one I found was to introduce some new functions in Bootstrap overriding the ones from Sass itself. The solution is available in #40925. It's only a draft to see the impacts so I put not so much efforts in testing and documentation.

That said, I still don't think it's a good solution to do it this way because:
- It's not the duty and goal of Bootstrap to maintain this kind of parsing functions.
- It's very probably a source a errors and problems for future us if we introduce this kind of code.
- It's an open door to every contribution wanting to add some new color spaces.
- What if Sass still deprecate some other functions, I think we won't be able to introduce these functions along the way.

>        2.1. **Drop `node-sass` support in a v5.3.x release.** This would be unfair to some users still relying on it, and frankly, I’m not a big fan of this option.

Definitely not a big fan too.

I can't decide if we should rather drop `node-sass` for 5.4 or for v6.

**Comment by julien-deramond** [comment]
My two cents on this.

I don't think it makes sense to release a v6 just to remove Node Sass. If that's part of v6, we'll be waiting a long time for the fix. Plus, there's no guarantee Sass won't release a v2 in the meantime, leaving us stuck on v1.78 for ages, just to support something that's been deprecated for quite a while.

I understand that removing Node Sass would be a breaking change, even in a minor version, so we'd need to finalize all pending patches—especially those related to color modes—within v5.3.x. After that, we could roll out a v5.4, merging everything that's ready, including dropping Node Sass support. Then, I imagine the next focus would shift to v6.

This way, Node Sass users wouldn’t lose significant features and still benefit from the v5 entirely, as they'd remain on v5.3.x instead of moving to v5.4, which likely wouldn't include any critical changes anyway, or big new features or improvements.

Thoughts @twbs/css-review, @mdo, folks, Node Sass users?

**Comment by ghiscoding** [comment]
@louismaximepiton @julien-deramond 
It might actually be possible to create a polyfill by creating a custom method that will check if the SASS function you're calling exists (`color.channel()`) if so use it or else use another SASS function (`lighten`). That is exactly what Angular-Material did in this old PR [fix warnings related to division operator](https://github.com/angular/components/pull/22871/files#diff-4b0d3692fae1e26970f5365a78db4ff81acb362f79ef12252081df8b285d38e8), so it might actually be possible to try this out (I didn't have time to create a new PR and I don't have, neither want to install, node-sass to try this out though, so...

Below was the Angular-Material polyfill for the old SASS `math.div()` warning, the project no longer has code since they released new major bumping SASS version but when this polyfill was implemented, nobody complained about it (they surely have thousands of users), so it's worth giving it try!?
```scss
@use 'sass:math';
@use 'sass:meta';
@use 'sass:list';

// Private polyfill for the `math.div` function from Sass to be used until we can update the
// minimum required Sass version to 1.34.0 or above.
// TODO(crisbeto): replace with `math.div` eventually.
@function private-div($a, $b) {
  @if (meta.function-exists('div', 'math')) {
    @return math.div($a, $b);
  }
  @else {
    @return $a / $b;
  }
}

// Private polyfill for the `list.slash` function from Sass to be used until we can update the
// minimum required Sass version to 1.34.0 or above.
// TODO(crisbeto): replace with `list.slash` eventually.
@function private-slash($a, $b) {
  @if (meta.function-exists('slash', 'list')) {
    @return list.slash($a, $b);
  }
  @else {
    @return #{$a}#{' / '}#{$b};
  }
}
```

Side note, I also just checked if node-sass already have the `meta.function-exists` and they do [here](https://github.com/search?q=repo%3Asass%2Fnode-sass%20function-exists&type=code)), so I'm assuming a similar polyfill for the new SASS function should work!

**Comment by louismaximepiton** [comment]
Hello, thanks for your contribution to the topic. 

We've been trying this feature and it sounds like an unsolvable issue. We are only human, so maybe we've been missing something in here.

Here are all our findings:
All the `@use "sass:**";` get in the generated CSS when it's generated via `node-sass`, not blocking but not very clean. Know that conditional `@import` or `@use` aren't possible.
It seems like we can't use `meta.function-exists` because it's not supported by `node-sass` so we need to use `function-exists` which only take one parameter so we can't test for functions inside modules (or we don't know the syntax). We couldn't find any function in Dart Sass not inside a Dart Sass module and not in Node Sass.

For Angular it worked because they were likely supporting 2 different versions of Dart Sass which is different and unfortunately not applicable to our issue.

**Comment by ghiscoding** [comment]
@louismaximepiton well ahh at least you tried :)
So in that case, I guess that v5.4.x would be a nice time to drop node-sass support then, I'm all in favor of that (though I guess it means that a fix is still a couple months away?). 
A big thanks for giving it a try.

**Comment by ntkme** [comment]
Just a heads-up that dart-sass is currently on track for emitting deprecation warnings for `@import` in upcoming `1.80.0` release: https://github.com/sass/dart-sass/pull/2282

**Comment by aepot** [comment]
@ntkme `import` warnings might be solved via bundler config and there's nothing to do with Bootstrap itself. For example, as per Vite documentation: https://vite.dev/config/shared-options#css-preprocessoroptions you may add `css` section to the Vite config with `api: 'modern-compiler'` option. I hope similar solution exists for Webpack.

**Comment by karolyi** [comment]
for webpack-dev-server, you can ignore warnings with the following configuration:

```js
  const devServerOptions = {
    hot: true,
    // host: '0.0.0.0',
    host: 'localhost',
    port: 3000,
    compress: true,
    historyApiFallback: true,
    allowedHosts: 'all',
    client: {
      overlay: {
        errors: true,
        runtimeErrors: true,
        warnings: false
      },
      progress: true,
    },
    headers: {
      'Access-Control-Allow-Origin': '*',
    },
    static: {
      directory: path.join(_myDirName, 'frontend', 'src', 'assets'),
    },
  }
```

the important parts are in the `client`->`overlay` section. this way, webpack will not flood your dev site with modals.

**Comment by mbabker** [comment]
In good news, dart-saas 1.80 also has options to silence these new deprecations, see https://sass-lang.com/blog/import-is-deprecated/#controlling-deprecation-warnings for more on that.

**Comment by evgeniy-vashchuk** [comment]
For everyone using Gulp who wants to temporarily hide these warnings, the is **silenceDeprecations** option for this:
```
.pipe(sass({
    silenceDeprecations: ['legacy-js-api', 'mixed-decls', 'color-functions', 'global-builtin', 'import'],
    outputStyle: config.isProd ? 'compressed' : 'expanded',
    indentType: 'space',
    indentWidth: 2,
    includePaths: ['./node_modules']
}).on('error', plugins.notify.onError({ title: 'Error compiling SASS' })))
```

**Comment by voltaek** [comment]
> I don't think it makes sense to release a v6 just to remove Node Sass. If that's part of v6, we'll be waiting a long time for the fix. Plus, there's no guarantee Sass won't release a v2 in the meantime, leaving us stuck on v1.78 for ages, just to support something that's been deprecated for quite a while.
> 
> I understand that removing Node Sass would be a breaking change, even in a minor version, so we'd need to finalize all pending patches—especially those related to color modes—within v5.3.x. After that, we could roll out a v5.4, merging everything that's ready, including dropping Node Sass support. Then, I imagine the next focus would shift to v6.

@julien-deramond It seems to me that the core of the issue here is everyone being stuck with this mindset that Bootstrap v6.0 has to be a "big upgrade" with new features and such, rather than v6.0 simply meaning that there was a breaking change and the version was incremented to reflect so, as per proper semantic versioning practices. Trying to neatly schedule when breaking changes happen is a nice idea in theory (and has been Bootstrap's modus operandi), but only really possible if your codebase exists is a vacuum.

If we are considering dropping `node-sass` support as a breaking change, then a major version increment should occur.

That being said, once that is done, I don't think that going from Dart Sass v1.xx to v2.0 (and even possibly to v3.0) would be considered a breaking change, so long as Bootstrap has already migrated to `@use` and handled all the other current deprecations e.g.: the color functions, since "breaking changes" are within the context of Bootstrap's source and the ability to build it with the tools it lists as dependencies.

**Comment by julien-deramond** [comment]
> > I don't think it makes sense to release a v6 just to remove Node Sass. If that's part of v6, we'll be waiting a long time for the fix. Plus, there's no guarantee Sass won't release a v2 in the meantime, leaving us stuck on v1.78 for ages, just to support something that's been deprecated for quite a while.
> > 
> > I understand that removing Node Sass would be a breaking change, even in a minor version, so we'd need to finalize all pending patches—especially those related to color modes—within v5.3.x. After that, we could roll out a v5.4, merging everything that's ready, including dropping Node Sass support. Then, I imagine the next focus would shift to v6.
> 
> @julien-deramond It seems to me that the core of the issue here is everyone being stuck with this mindset that Bootstrap v6.0 has to be a "big upgrade" with new features and such, rather than v6.0 simply meaning that there was a breaking change and the version was incremented to reflect so, as per proper semantic versioning practices. Trying to neatly schedule when breaking changes happen is a nice idea in theory (and has been Bootstrap's modus operandi), but only really possible if your codebase exists is a vacuum.
> 
> If we are considering dropping `node-sass` support as a breaking change, then a major version increment should occur.
> 
> That being said, once that is done, I don't think that going from Dart Sass v1.xx to v2.0 (and even possibly to v3.0) would be considered a breaking change, so long as Bootstrap has already migrated to `@use` and handled all the other current deprecations e.g.: the color functions, since "breaking changes" are within the context of Bootstrap's source and the ability to build it with the tools it lists as dependencies.

Personally, I agree that we should release more often, even major versions. However, since it's not been decided yet to do so, I've suggested that this breaking change (dropping `node-sass` + migration to color functions and `@use`) could be tackled in the last version of Bootstrap 5, so a 5.5. I've created a draft PR reflecting that.

**Comment by stefanahman** [comment]
Here's the PR for anyone wondering: https://github.com/twbs/bootstrap/pull/41112

**Comment by StubbornDeer** [comment]
> Here's the PR for anyone wondering: [#41112](https://github.com/twbs/bootstrap/pull/41112)

Thanks for this update but can you please explain how to use this patch? Bumped into this problem today :(

**Comment by stefanahman** [comment]
> Thanks for this update but can you please explain how to use this patch? Bumped into this problem today :(

There seem to be no patch yet, as the PR is in draft. It think if we practice some patience, it will be ready before the deprecations will turn into errors.

**Comment by Berzerk3000** [comment]
> > Thanks for this update but can you please explain how to use this patch? Bumped into this problem today :(
> 
> There seem to be no patch yet, as the PR is in draft. It think if we practice some patience, it will be ready before the deprecations will turn into errors.

bootstrap will NOT fix the issue. they even rejected a PR with fix for this issue afaik. I believe if you want to use bootstrap with sass you are stuck at sass 1.77.4 or so.

**Comment by aepot** [comment]
@Berzerk3000 then it's time to drop it from dependencies i guess.

**Comment by julien-deramond** [comment]
> bootstrap will NOT fix the issue. they even rejected a PR with fix for this issue afaik. 

Please refer to the active PR at https://github.com/twbs/bootstrap/pull/41112 linked to this issue. It's important to note that while the changes may seem straightforward, merging this PR requires careful consideration and testing due to the complexity of the Bootstrap project and retro-compatibility with Node Sass. Moreover, as a volunteer open-source contributor, my time is limited by personal commitments and other responsibilities. I'll try to do my best 🙏

**Comment by C-Ezra-M** [comment]
Bootstrap doesn't use `@use`, so you must use `@import` to import Bootstrap in Sass. You can't use `@use`, otherwise you get errors like "undefined mixin". It has worked before because `@import` puts everything imported into the global scope. Numerous Bootstrap files don't even use `@import`!

`@import` is deprecated, and eventually it will simply resolve to a plain CSS `@import`. https://sass-lang.com/documentation/breaking-changes/import/
